sdk/js/src/svm.ts

```diff
@@ -82,6 +88,175 @@ export async function constructSubmitBidInstruction(
+
+export function createAssociatedTokenAccountIdempotentInstruction(
+  owner: PublicKey,
+  mint: PublicKey,
+  payer: PublicKey = owner,
+  tokenProgram: PublicKey
+): [PublicKey, TransactionInstruction] {
+  const ataAddress = getAssociatedTokenAddress(owner, mint, tokenProgram);
+  const createUserTokenAccountIx = createAssociatedTokenAccountInstruction(
+    payer,
+    ataAddress,
+    owner,
+    mint,
+    tokenProgram,
+    ASSOCIATED_TOKEN_PROGRAM_ID
+  );
+  // idempotent ix discriminator is 1
+  createUserTokenAccountIx.data = Buffer.from([1]);
+  return [ataAddress, createUserTokenAccountIx];
+}
+
+export async function constructSwapBid(
+  tx: Transaction,
+  searcher: PublicKey,
+  swapOpportunity: OpportunitySvmSwap,
+  bidAmount: anchor.BN,
+  deadline: anchor.BN,
+  chainId: string,
+  relayerSigner: PublicKey
+): Promise<BidSvm> {
+  const expressRelay = new Program<ExpressRelay>(
+    expressRelayIdl as ExpressRelay,
+    {} as AnchorProvider
+  );
+  const expressRelayMetadata = getExpressRelayMetadataPda(chainId);
+  const svmConstants = SVM_CONSTANTS[chainId];
+
+  const tokenProgramInput = TOKEN_PROGRAM_ID;
+  const tokenProgramOutput = TOKEN_PROGRAM_ID;
+  const tokenProgramFee = TOKEN_PROGRAM_ID;
+  const mintInput =
+    swapOpportunity.tokens.type === "input_specified"
+      ? swapOpportunity.tokens.inputToken.token
+      : swapOpportunity.tokens.inputToken;
+  const mintOutput =
+    swapOpportunity.tokens.type === "output_specified"
+      ? swapOpportunity.tokens.outputToken.token
+      : swapOpportunity.tokens.outputToken;
+  const trader = swapOpportunity.userWalletAddress;
+  const mintFee = mintInput;
+  const router = swapOpportunity.routerAccount;
+
+  const swapArgs = {
+    amountInput:
+      swapOpportunity.tokens.type === "input_specified"
+        ? new anchor.BN(swapOpportunity.tokens.inputToken.amount)
+        : bidAmount,
+    amountOutput:
+      swapOpportunity.tokens.type === "output_specified"
+        ? new anchor.BN(swapOpportunity.tokens.outputToken.amount)
+        : bidAmount,
+    referralFeeBps: new anchor.BN(0),
+    deadline,
+    feeToken: { input: {} },
+  };
+  const ixSwap = await expressRelay.methods
+    .swap(swapArgs)
+    .accountsStrict({
+      expressRelayMetadata,
+      searcher,
+      trader: swapOpportunity.userWalletAddress,
+      searcherInputTa: getAssociatedTokenAddress(
+        searcher,
+        mintInput,
+        tokenProgramInput
+      ),
+      searcherOutputTa: getAssociatedTokenAddress(
+        searcher,
+        mintOutput,
+        tokenProgramOutput
+      ),
+      traderInputAta: getAssociatedTokenAddress(
+        trader,
+        mintInput,
+        tokenProgramInput
+      ),
+      tokenProgramInput,
+      mintInput,
+      traderOutputAta: getAssociatedTokenAddress(
+        trader,
+        mintOutput,
+        tokenProgramOutput
+      ),
+      tokenProgramOutput,
+      mintOutput,
+      routerFeeReceiverTa: getAssociatedTokenAddress(
```

Comment:
this is not necessarily the ATA. the router account here will be the token account that the referrer wants the SPL fee token to be sent to, so you can just use that